### PR TITLE
#395: fields with no default value/factory should be marked as required when generating schema

### PR DESCRIPTION
**Pull request by deansg**
When generating a marshmallow schema for a dataclass, fields that have no default value or default factory should be marked as required, so that when deserializing an object into the dataclass, a missing value for the field will result in a proper Validation error instead of a KeyError.

### CODE REVIEW

File: dataclasses_json/mm.py [hunk 1]
```diff
@@ -282,18 +282,19 @@ def schema(cls, mixin, infer_missing):
...
     #  https://marshmallow.readthedocs.io/en/stable/quickstart.html
     for field in dc_fields(cls):
-        metadata = (field.metadata or {}).get('dataclasses_json', {})
         metadata = overrides[field.name]
         if metadata.mm_field is not None:
             schema[field.name] = metadata.mm_field
         else:
             type_ = field.type
-            options = {}
+            options: typing.Dict[str, any] = {}
```
Comment (george-zubrienko) [inline dataclasses_json/mm.py:290]: Should be `Any`?

**Review by george-zubrienko — COMMENTED** [review]
Hi @deansg! I see that CI fails - could you take a look and try fixing it?

File: dataclasses_json/mm.py (hunk 1 above)
Comment (deansg) [reply dataclasses_json/mm.py]: Fixed

**Comment by deansg** [comment]
@george-zubrienko Hi George, I fixed the build, but had to basically override the expected behavior in 2 tests for this to pass. I feel like the new behavior is a lot more correct, but technically this can be considered breaking.

**Comment by george-zubrienko** [comment]
@lidatong will you have time to make a decision here? I'd like you to sign off either the merge as a regular patch, or as a breaking since test expectations are changed

**Comment by deansg** [comment]
Basically my logic is: if you can't instantiate it as a Python object with an empty dictionary, then an empty JSON should also fail to be converted.
This is also the logic in the marshmallow_dataclass library, which today I use in addition to dataclass_json

**Comment by george-zubrienko** [comment]
I believe this is a correct change, but I'll wait couple days more in case others have different opinions. If nothing comes in, I'll merge this as a bugfix around June 21

**Comment by lidatong** [comment]
> @lidatong will you have time to make a decision here? I'd like you to sign off either the merge as a regular patch, or as a breaking since test expectations are changed

yes this makes sense to me as a fix

**Review by lidatong — DISMISSED** [review]

**Comment by deansg** [comment]
I see there's a conflict there that has an easy solution, do you want me to resolve it or is it better a maintainer do it since they are the ones that approve/merge? @george-zubrienko

**Comment by george-zubrienko** [comment]
> I see there's a conflict there that has an easy solution, do you want me to resolve it or is it better a maintainer do it since they are the ones that approve/merge? @george-zubrienko 

@deansg Please go ahead and resolve conflicts for you PR, will merge right away

**Comment by deansg** [comment]
@george-zubrienko resolved, seem like some styling check is failing now but it mentions a lot of errors unrelated to my PRs, maybe it's a new check?

**Review by george-zubrienko — APPROVED** [review]

**Comment by george-zubrienko** [comment]
> @george-zubrienko resolved, seem like some styling check is failing now but it mentions a lot of errors unrelated to my PRs, maybe it's a new check?

Its all good, this check is optional and we'll make it green soon. You can help if you have time :)

**Pull request merged**